### ISSUE TITLE
Optionally implement terraform-provider-jinja as provider-defined function(s) for Terraform 1.8+

### ISSUE THREAD

**Issue by crw** [NikolaLohinski/terraform-provider-jinja#15]
Thank you for supporting the Terraform community with this data provider! If you were using the data provider construct to work around the lack of support in Terraform for external functions, please consider the following new feature of Terraform. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement this functionality as a function, rather than wrap the functionality in a data provider. Provider-based functions have the advantage of a simpler in-line syntax that does not require the data block (or any other resource blocks) to use.

The caveat is that, as provider-defined functions are new to Terraform 1.8, you would need to continue to maintain the current data provider functionality if you wished to continue to support earlier versions of Terraform. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you have any questions, please visit the [Terraform topic in our official forum](https://discuss.hashicorp.com/c/terraform-core/27). 

We greatly appreciate your support of the Terraform community with your contributions. **If you do not wish to implement this functionality as a function, please feel free to close this issue**. This issue is simply to raise awareness of the new capability. Thanks again!

**Comment by NikolaLohinski** [comment]
Hi @crw and thanks for the suggestion.

I'll keep the issue opened and see if I can give it a try some time. 

Given the amount of possible configuration parameters, I don't think a function would look as clean as a data source block, but it could be a simplified helper function with some predefined options.

For anyone reading this and feeling motivated, I am happy to take PRs.

Cheers

**Comment by jamesgeddes** [comment]
FWIW, I like:
- the data block,
- backward compatibility,
- not having to maintain two versions of the same functionality.

As you mention, @NikolaLohinski, I am not sure how switching to a provider-defined function would improve the use case I currently have. Here is what I am doing with the [jinja_template data block](https://github.com/evoteum/tofu-modules/blob/main/github/workflows/github_files.tofu#L23-L46):
```hcl
data "jinja_template" "workflow" {
  for_each = local.enabled_workflows

  context {
    type = "json"
    data = jsonencode({
      workflow_name            = each.key
      controller_repo_name     = var.controller_repo_name
      cron_description         = lookup(local.workflow_cron_description, each.key, "Custom schedule")
      cron_expression          = lookup(local.workflow_cron_expression, each.key, null)
      organisation             = var.organisation
      repository_configuration = local.effective_repository_configuration[each.key]
      reusable_workflows_repo  = var.reusable_workflows_repository
      source_path              = lookup(var.repository_configuration, "source_path", "*")
      workflow_inputs          = local.workflow_inputs[each.key]
      workflow_permissions     = local.workflow_permissions[each.key]
    })
  }

  source {
    template  = file("${path.module}/templates/reusable-workflow-caller.yml.j2")
    directory = path.module
  }
}
```

If this were implemented as a provider-defined function, if I have understood correctly, it might look something like this:
```hcl
output "workflow_templates" {
  value = {
    for key, cfg in local.enabled_workflows : key => jinja.render(
      template  = file("${path.module}/templates/reusable-workflow-caller.yml.j2"),
      context   = {
        type = "json"
        data = jsonencode({
          workflow_name            = key
          controller_repo_name     = var.controller_repo_name
          cron_description         = lookup(local.workflow_cron_description, key, "Custom schedule")
          cron_expression          = lookup(local.workflow_cron_expression, key, null)
          organisation             = var.organisation
          repository_configuration = local.effective_repository_configuration[key]
          reusable_workflows_repo  = var.reusable_workflows_repository
          source_path              = lookup(var.repository_configuration, "source_path", "*")
          workflow_inputs          = local.workflow_inputs[key]
          workflow_permissions     = local.workflow_permissions[key]
        })
      },
      directory = path.module
    )
  }
}
```

However, that version arguably introduces a number of downsides:
- The data block is declarative, readable, and testable. The function version becomes a deeply nested expression that is harder to parse and maintain.
- The loop still exists, but now lives inside a large expression, reducing clarity.
- With the data block, I can inspect individual items like `data.jinja_template.workflow["example"]` easily in state and plans. Would the function version be just be a blob of text in an output map?
- Having to pass `directory = path.module` into a function also feels awkward; this is a detail better suited to a block-based configuration.

In short, the provider-defined function would not simplify my use case and would arguably make it harder to read, debug, and maintain. I can see how this functionality might add value for other providers with smaller or more dynamic expressions, but for rendering complex templates from static files, the current data source model feels ideal.

![Image](https://github.com/user-attachments/assets/c05d79c5-fe3f-43ad-a1fa-4c05c4573b8b)
